control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -211,33 +173,120 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
-	if err != nil {
-		return ctrl.Result{}, err
+	// Ensure the hostedControlPlane has a finalizer for cleanup
+	if !controllerutil.ContainsFinalizer(hostedControlPlane, finalizer) {
+		controllerutil.AddFinalizer(hostedControlPlane, finalizer)
+		if err := r.Update(ctx, hostedControlPlane); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer to hostedControlPlane: %w", err)
+		}
 	}
 
-	if cluster == nil {
-		r.Log.Info("Cluster Controller has not yet set OwnerRef")
-		return ctrl.Result{}, nil
+	// Reconcile etcd cluster status
+	{
+		newCondition := metav1.Condition{
+			Type:   string(hyperv1.EtcdAvailable),
+			Status: metav1.ConditionUnknown,
+			Reason: "EtcdStatusUnknown",
+		}
+		etcdCluster := manifests.EtcdCluster(hostedControlPlane.Namespace)
+		if err := r.Get(ctx, types.NamespacedName{Namespace: etcdCluster.Namespace, Name: etcdCluster.Name}, etcdCluster); err != nil {
+			if apierrors.IsNotFound(err) {
+				newCondition = metav1.Condition{
+					Type:   string(hyperv1.EtcdAvailable),
+					Status: metav1.ConditionFalse,
+					Reason: "EtcdClusterNotFound",
+				}
+			} else {
+				return ctrl.Result{}, fmt.Errorf("failed to fetch etcd cluster %s/%s: %w", etcdCluster.Namespace, etcdCluster.Name, err)
+			}
+		} else {
+			cond, err := etcd.ComputeEtcdClusterStatus(ctx, r.Client, etcdCluster)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to compute etcd cluster status: %w", err)
+			}
+			newCondition = cond
+		}
+		newCondition.ObservedGeneration = hostedControlPlane.Generation
+		meta.SetStatusCondition(&hostedControlPlane.Status.Conditions, newCondition)
 	}
 
-	if util.IsPaused(cluster, hostedControlPlane) {
-		r.Log.Info("HostedControlPlane or linked Cluster is marked as paused. Won't reconcile")
-		return ctrl.Result{}, nil
+	// Reconcile Kube APIServer status
+	{
+		newCondition := metav1.Condition{
+			Type:   string(hyperv1.KubeAPIServerAvailable),
+			Status: metav1.ConditionUnknown,
+			Reason: "StatusUnknown",
+		}
+		deployment := manifests.KASDeployment(hostedControlPlane.Namespace)
+		if err := r.Get(ctx, client.ObjectKeyFromObject(deployment), deployment); err != nil {
+			if apierrors.IsNotFound(err) {
+				newCondition = metav1.Condition{
+					Type:   string(hyperv1.KubeAPIServerAvailable),
+					Status: metav1.ConditionFalse,
+					Reason: "DeploymentNotFound",
+				}
+			} else {
+				return ctrl.Result{}, fmt.Errorf("failed to fetch Kube APIServer deployment %s/%s: %w", deployment.Namespace, deployment.Name, err)
+			}
+		} else {
+			// Assume the deployment is unavailable until proven otherwise.
+			newCondition = metav1.Condition{
+				Type:   string(hyperv1.KubeAPIServerAvailable),
+				Status: metav1.ConditionFalse,
+				Reason: "DeploymentStatusUnknown",
+			}
+			for _, cond := range deployment.Status.Conditions {
+				if cond.Type == appsv1.DeploymentAvailable && cond.Status == corev1.ConditionTrue {
```

Comment:
This check is not enough. You also need https://github.com/openshift/hypershift/blob/c712bc64482f6b57f08c5754d8c1d7b9086e02c7/control-plane-operator/controllers/hostedcontrolplane/kas/status.go#L23
The DeploymentAvailable condition is set to true as soon as there is a replica regardless of whether that replica is available or not.

Reply:
Video of deployment status as it becomes available:
https://drive.google.com/file/d/15c9WgCJQKFYFE9H-L82lxiCe_6VbeYsM/view?usp=sharing

Reply:
If the deployment reports `Available=True` but isn't actually available, to me that sounds like a misconfigured deployment or missing/flaky readiness checks for the pods. The point of the condition is to abstract the computation for availability from consumers to avoid, for example, rewriting the k8s deployment controller logic in our own code:

https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

>Conditions should complement more detailed information about the observed status of an object written by a controller, rather than replace it. For example, the "Available" condition of a Deployment can be determined by examining readyReplicas, replicas, and other properties of the Deployment. However, the "Available" condition allows other components to avoid duplicating the availability logic in the Deployment controller.

And in fact I don't see any liveness or readiness probes defined here: https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kas/deployment.go#L110 — did I miss them?

Reply:
The liveness and readiness probes are defined here:
https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/kas/params.go#L153-L178
They are configurable

Reply:
I checked the deployment and both liveness and readiness probes are there and match what we have in standalone OCP

Reply:
And the deployment's rolling update strategy is configured with MaxUnavailable=1. Doesn't that mean the rollout will tolerate 0 available replicas for the Available condition?

I would expect us to be using MaxUnavailable=0 and MaxSurge=1 by default.

Reply:
Aha, let me try that

Reply:
Yup, that does it. Now Available=true means that there is a replica that is available. Will you make that change in this PR?

Reply:
Also, should the MaxUnavailable be different depending on whether we have 1 vs 3 replicas?

Reply:
I changed MaxUnavailable=0 for this PR and we can take the config topic separately (shouldn't 3 should be good enough to handle both use cases in the meantime?)